cpu/stm32l1/periph/pm.c

```diff
@@ -0,0 +1,359 @@
+{
+    return (GPIO_TypeDef *)(pin & ~(0x0f));
+}
+
+/**
+ * @brief   Extract the port number form the given identifier
+ *
+ * The port number is extracted by looking at bits 10, 11, 12, 13 of the base
+ * register addresses.
+ */
+static inline int _port_num(gpio_t pin)
+{
+    return ((pin >> 10) & 0x0f);
+}
+
+/**
+ * @brief   Extract the pin number from the last 4 bit of the pin identifier
+ */
+static inline int _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+/* We are not using gpio_init as it sets GPIO clock speed to maximum */
+/* We add GPIOs we touched to exclusion mask pm_portmask_system */
+static void pin_set(gpio_t pin, uint8_t value) {
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
+    tmpreg = port->MODER;
+    tmpreg &= ~(3 << (2*pin_num));
+    tmpreg |= (1 << (2*pin_num));
+    port->MODER = tmpreg;
+
+    port->PUPDR &= ~(3 << (2*pin_num));
+    port->OTYPER &= ~(1 << pin_num);
+    if (value) {
+        port->ODR |= (1 << pin_num);
+    } else {
+        port->ODR &= ~(1 << pin_num);
+    }
+
+    pm_portmask_system[_port_num(pin)] |= 1 << pin_num;
+}
+
+/* Do not change GPIO state in sleep mode */
+void pm_arch_add_gpio_exclusion(gpio_t gpio) {
+    uint8_t port = _port_num(gpio);
+    uint8_t pin = _pin_num(gpio);
+
+    pm_portmask_user[port] |= (uint16_t)(1<<pin);
+}
+
+/* Change GPIO state to AIN in sleep mode */
+void pm_arch_del_gpio_exclusion(gpio_t gpio) {
+    uint8_t port = _port_num(gpio);
+    uint8_t pin = _pin_num(gpio);
+
+    pm_portmask_user[port] &= ~(uint16_t)(1<<pin);
+}
+
+void pm_before_i_go_to_sleep(void){
+    uint8_t i;
+    uint8_t p;
+    uint32_t mask;
+    GPIO_TypeDef *port;
+
+
+    for (i = 0; i < CPU_NUMBER_OF_PORTS; i++) {
+        port = (GPIO_TypeDef *)(GPIOA_BASE + i*(GPIOB_BASE - GPIOA_BASE));
+
+        /* save GPIO registers values */
+        pm_gpio_moder[i] = port->MODER;
+        pm_gpio_pupdr[i] = port->PUPDR;
+        pm_gpio_otyper[i] = (uint16_t)(port->OTYPER & 0xFFFF);
+        pm_gpio_odr[i] = (uint16_t)(port->ODR & 0xFFFF);
+    }
+
+    /* Disable all USART interfaces in use */
+    /* without it, RX will receive some garbage when MODER is changed */
+
+    for (i = 0; i < UART_NUMOF; i++) {
+        if (uart_config[i].dev->CR1 & USART_CR1_UE) {
+            uart_config[i].dev->CR1 &= ~USART_CR1_UE;
+            pin_set(uart_config[i].tx_pin, 1);
+            pm_usart[i] = 1;
+        } else {
+            pm_usart[i] = 0;
+        }
+    }
+
+    /* specifically set GPIOs used for external SPI devices */
+    /* NSS = 1, MOSI = 0, SCK = 0, MISO doesn't matter */
+    for (i = 0; i < SPI_NUMOF; i++) {
+        /* port and pin numbers */
+        GPIO_TypeDef *port = _port(spi_config[i].sclk_pin);
+        int pin_num = _pin_num(spi_config[i].sclk_pin);
+
+        /* check if alternate functions are enabled for that device*/
+        if (((port->AFR[(pin_num > 7) ? 1 : 0] >> (pin_num * 4)) & 0x0f)  == spi_config[i].af) {
```

Comment:
This line is very confusing. Can you factor it out a bit?

Reply:
This one is to check that SPI port is actually in use (corresponding pin is in AF mode) before reconfiguring it.

But the more I think of it — the more I tend to agree with you the code is too complicated and mostly unnecessary to achieve low power.

Reply:
Thanks for the explanation! But what I meant is that there is an usage of the ternary operator inside an `if` statement, which makes the whole line a bit complicated to read, plus the `4` which I don't really know what it means.

Regarding the function, I wouldn't say it's unnecessary, the function can be very helpful to configure the low power mode on this peripheral but what it's overhead is to do it systematically, regardless of the current configuration a user might have done.

Reply:
My point is, there is actually no need to configure peripherals for low power on STM32. Clocks are stopped in STOP mode, peripheral controllers are stopped, and GPIOs will preserve the state they were left in.

This function was to set SPI GPIO state, which is unnecessary as the same state should be already set by SPI controller itself, and it's not changing when entering STOP or SLEEP.

So the code can be removed.

Reply:
Ok I got it, thanks again for the clarification!